pynisher/pynisher.py

```diff
@@ -0,0 +1,450 @@
+
+    pass
+
+
+EMPTY = _EMPTY()
+
+
+class Pynisher(ContextDecorator):
+    """Restrict a function's resources"""
+
+    def __init__(
+        self,
+        func: Callable,
+        *,
+        name: str | None = None,
+        memory: int | tuple[int, str] | None = None,
+        cpu_time: int | tuple[float, str] | None = None,
+        wall_time: int | tuple[float, str] | None = None,
+        context: str | None = None,
+        raises: bool = True,
+        warnings: bool = True,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        func : Callable
+            The function to limit and call
+
+        name : str | None
+            A name to give the process that gets created, defaults to whatever
+            multiprocessing.Process defaults to.
+
+        memory : int | tuple[int, str] | None = None
+            The amount of memory to limit by. If `tuple`, specify units with (4, "MB").
+            Possible units are "B", "KB", "MB", "GB".
+
+            Processes take up some space before any limitation can take place.
+            These will run fine until a new allocation takes place.
+            This means a process can technically run in a limit of 1 Byte, up until the
+            point it tries to allocate anything.
+
+        cpu_time : int | tuple[float, str] | None = None
+            The cpu time in seconds to limit the process to. This time is only counted while the
+            process is active.
+            Can provide in (time, units) such as (1.5, "h") to indicate one and a half hours.
+            Units available are "s", "m", "h".
+
+        wall_time : int | tuple[float, str] | None = None
+            The amount of total wall time in seconds to limit to
+            Can provide in (time, units) such as (1.5, "h") to indicate one and a half hours.
+            Units available are "s", "m", "h"
+
+        context : "fork" | "forkserver" | "spawn" | None = None
+            The context to use with multiprocessing.get_context()
+            * https://docs.python.org/3/library/multiprocessing.html#multiprocessing.get_context
+
+        raises : bool = True
+            Whether any error from the subprocess should filter up and be raised.
+
+        warnings : bool = True
+            Whether to emit pynisher warnings or not.
+        """  # noqa
+        if isinstance(cpu_time, tuple):
+            x, unit = cpu_time
+            cpu_time = round(timeconvert(x, frm=unit))
+
+        if isinstance(wall_time, tuple):
+            x, unit = wall_time
+            wall_time = round(timeconvert(x, frm=unit))
+
+        if isinstance(memory, tuple):
+            x, unit = memory
+            memory = round(memconvert(x, frm=unit))
+
+        if not callable(func):
+            raise ValueError(f"`func` ({func}) must be callable")
+
+        if cpu_time is not None and not cpu_time >= 1:
+            raise ValueError(f"`cpu_time` ({cpu_time}) must be >= 1 seconds")
```

Comment:
As all of `cpu_time`, `wall_time`, `memory` and `context` are converted, it could be a good idea to also print the converted value in case of an error message.

Reply:
Yup, will fix